File: src/systems/CraftingSystem.ts
// CraftingSystem: the active artisan loop for Cooking / Smithing / Carpentry —
// tick-based like gathering, with item mastery (preserve chance + speed) per
// GDD §5.A. Building-gated recipes (Smithing/Carpentry) check hasBuilding().
import type { GameState } from "../state/GameState";
import type { CraftRecipe } from "../data/Recipes";
import { addItem, isBulk, removeItem, countItem, storedAmount } from "../components/Inventory";
import { addMasteryXp, masteryLevel } from "../components/Skills";
import { levelFromXp } from "../data/XPTable";
import { TICK_MS } from "../core/Engine";
import type { BuildingType } from "../data/Buildings";

export interface CraftEvent {
  recipe: CraftRecipe;
  amount: number;
  xpGained: number;
  preserved: boolean;
  burned: boolean;
}

export type CraftEndReason = "stopped" | "level_shortfall" | "missing_materials" | "missing_building" | "inventory_full";

export interface CraftCallbacks {
  onCraft?: (e: CraftEvent) => void;
  onStart?: (r: CraftRecipe) => void;
  onEnd?: (r: CraftRecipe | null, reason: CraftEndReason) => void;
}

export class CraftingSystem {
  private state: GameState;
  private hasBuilding: (type: BuildingType) => boolean;
  private cb: CraftCallbacks = {};
  private active: CraftRecipe | null = null;
  private tickAcc = 0;
  private ticksNeeded = 0;

  constructor(state: GameState, hasBuilding: (type: BuildingType) => boolean) {
    this.state = state;
    this.hasBuilding = hasBuilding;
  }

  setCallbacks(cb: CraftCallbacks) { this.cb = cb; }
  get activeRecipe() { return this.active; }
  get hasActive() { return this.active !== null; }
  get progress() {
    if (!this.active) return 0;
    return Math.min(1, this.tickAcc / (this.ticksNeeded * TICK_MS));
  }

  /** Why `recipe` can't run right now, or null if it's ready. */
  canStart(recipe: CraftRecipe): CraftEndReason | null {
    const lvl = levelFromXp(this.state.player.skills[recipe.skill].xp);
    if (lvl < recipe.levelReq) return "level_shortfall";
    if (recipe.requiresBuilding && !this.hasBuilding(recipe.requiresBuilding)) return "missing_building";
    for (const inp of recipe.inputs) {
      if (countItem(this.state.player.inventory, inp.itemId) < inp.qty) return "missing_materials";
    }
    return null;
  }

  start(recipe: CraftRecipe): boolean {
    const reason = this.canStart(recipe);
    if (reason) { this.cb.onEnd?.(recipe, reason); return false; }
    this.active = recipe;
    this.tickAcc = 0;
    this.ticksNeeded = this.actionTicks(recipe);
    this.cb.onStart?.(recipe);
    return true;
  }

  stop(reason: CraftEndReason = "stopped") {
    const r = this.active;
    this.active = null;
    if (r) this.cb.onEnd?.(r, reason);
  }

  private actionTicks(recipe: CraftRecipe): number {
    const m = this.masteryLevelFor(recipe);
    const frac = Math.min(1, m / 99);
    const floor = Math.max(2, Math.ceil(recipe.ticks * 0.6));
    let ticks = Math.max(floor, Math.round(recipe.ticks * (1 - frac * 0.33)));
    // P3.4: a built Campfire makes cooking 25% faster.
    if (recipe.skill === "cooking" && this.hasBuilding("CAMPFIRE")) {
      ticks = Math.max(2, Math.round(ticks * 0.75));
    }
    return ticks;
  }

  private masteryLevelFor(recipe: CraftRecipe): number {
    const xp = this.state.player.skills[recipe.skill].mastery[recipe.id] || 0;
    return masteryLevel(xp);
  }

  tick(dtMs: number) {
    if (!this.active) return;
    this.tickAcc += dtMs;
    if (this.tickAcc < this.ticksNeeded * TICK_MS) return;
    this.tickAcc = 0;
    this.performCraft(this.active);
  }

  private performCraft(recipe: CraftRecipe) {
    const reason = this.canStart(recipe);
    if (reason) { this.stop(reason); return; }
    const inv = this.state.player.inventory;

    const mLvl = this.masteryLevelFor(recipe);
    const preserveChance = Math.min(0.15, (mLvl / 99) * 0.15);
    const preserved = Math.random() < preserveChance;

    let burned = false;
    if (recipe.burnable) {
      const lvl = levelFromXp(this.state.player.skills[recipe.skill].xp);
      const margin = Math.max(0, lvl - recipe.levelReq);
      burned = Math.random() < Math.max(0, 0.35 - margin * 0.02 - (mLvl / 99) * 0.1);
    }

    if (!preserved) {
      for (const inp of recipe.inputs) removeItem(inv, inp.itemId, inp.qty);
    }

    let xpGained: number;
    let amount = 0;
    if (!burned) {
      const out = recipe.output.itemId;
      if (isBulk(out) && storedAmount(inv) + recipe.output.qty > inv.storageCap) { this.stop("inventory_full"); return; }
      addItem(inv, recipe.output.itemId, recipe.output.qty);
      amount = recipe.output.qty;
      xpGained = recipe.xp;
      this.state.collectionLog.add(recipe.output.itemId);
    } else {
      xpGained = Math.round(recipe.xp * 0.2);
    }
    this.state.player.skills[recipe.skill].xp += xpGained;
    addMasteryXp(this.state.player.skills, recipe.skill, recipe.id, Math.max(1, recipe.output.qty) * 4);

    this.cb.onCraft?.({ recipe, amount, xpGained, preserved, burned });

    // Loop: keep crafting the same recipe while materials/level allow.
    const next = this.canStart(recipe);
    if (next) this.stop(next);
  }
}

File: unity/Assets/Isoperia/Core/Runtime/Systems/CraftingSystem.cs
using System;
using Isoperia.Core.Components;
using Isoperia.Core.Content;
using Isoperia.Core.Data;
using Isoperia.Core.Save;
using Isoperia.Core.Sim;
using Isoperia.Core.State;

namespace Isoperia.Core.Systems
{
    public enum CraftEndReason { Stopped, LevelShortfall, MissingMaterials, MissingBuilding, InventoryFull }

    public sealed class CraftEvent
    {
        public JsonValue Recipe;
        public int Amount;
        public int XpGained;
        public bool Preserved;
        public bool Burned;
    }

    /// <summary>Core tick-driven recipe loop, ported from the web CraftingSystem.</summary>
    public sealed class CraftingSystem
    {
        private readonly GameState state;
        private readonly ContentDatabase content;
        private readonly IRandom random;
        private readonly Func<string, bool> hasBuilding;
        private JsonValue active;
        private double elapsed;
        private int ticksNeeded;

        public event Action<JsonValue> Started;
        public event Action<CraftEvent> Crafted;
        public event Action<JsonValue, CraftEndReason> Ended;
        public JsonValue ActiveRecipe => active;

        public CraftingSystem(GameState state, ContentDatabase content, IRandom random, Func<string, bool> hasBuilding)
        {
            this.state = state ?? throw new ArgumentNullException(nameof(state));
            this.content = content ?? throw new ArgumentNullException(nameof(content));
            this.random = random ?? throw new ArgumentNullException(nameof(random));
            this.hasBuilding = hasBuilding ?? throw new ArgumentNullException(nameof(hasBuilding));
        }

        public CraftEndReason? CanStart(JsonValue recipe)
        {
            string skill = recipe["skill"].AsString();
            if (state.Player.Skills.LevelOf(skill) < (int)recipe["levelReq"].AsNumber(1)) return CraftEndReason.LevelShortfall;
            string building = recipe["requiresBuilding"].AsString();
            if (!string.IsNullOrEmpty(building) && !hasBuilding(building)) return CraftEndReason.MissingBuilding;
            foreach (JsonValue input in recipe["inputs"].Items)
                if (state.Player.Inventory.Count(input["itemId"].AsString()) < (int)input["qty"].AsNumber()) return CraftEndReason.MissingMaterials;
            return null;
        }

        public bool Start(JsonValue recipe)
        {
            CraftEndReason? reason = CanStart(recipe);
            if (reason.HasValue) { Ended?.Invoke(recipe, reason.Value); return false; }
            active = recipe;
            elapsed = 0;
            ticksNeeded = ActionTicks(recipe);
            Started?.Invoke(recipe);
            return true;
        }

        public void Stop(CraftEndReason reason = CraftEndReason.Stopped)
        {
            JsonValue recipe = active;
            active = null;
            if (recipe != null) Ended?.Invoke(recipe, reason);
        }

        public void Tick(double dtMs)
        {
            if (active == null) return;
            elapsed += dtMs;
            if (elapsed < ticksNeeded * TickRunner.TickMs) return;
            elapsed = 0;
            Perform(active);
        }

        private int ActionTicks(JsonValue recipe)
        {
            string skill = recipe["skill"].AsString();
            state.Player.Skills.Get(skill).Mastery.TryGetValue(recipe["id"].AsString(), out double mastery);
            int masteryLevel = MasteryTable.LevelFromXp(mastery);
            double baseTicks = recipe["ticks"].AsNumber(1);
            int ticks = (int)Math.Max(Math.Max(2, Math.Ceiling(baseTicks * .6)), Math.Round(baseTicks * (1 - Math.Min(1, masteryLevel / 99d) * .33)));
            return skill == Skills.Cooking && hasBuilding("CAMPFIRE") ? Math.Max(2, (int)Math.Round(ticks * .75)) : ticks;
        }

        private void Perform(JsonValue recipe)
        {
            CraftEndReason? gate = CanStart(recipe);
            if (gate.HasValue) { Stop(gate.Value); return; }
            string skill = recipe["skill"].AsString();
            state.Player.Skills.Get(skill).Mastery.TryGetValue(recipe["id"].AsString(), out double mastery);
            int masteryLevel = MasteryTable.LevelFromXp(mastery);
            bool preserved = random.Next() < Math.Min(.15, masteryLevel / 99d * .15);
            bool burned = recipe["burnable"].AsBool(false) && random.Next() < Math.Max(0, .35 - Math.Max(0, state.Player.Skills.LevelOf(skill) - (int)recipe["levelReq"].AsNumber()) * .02 - masteryLevel / 99d * .1);
            if (!preserved) foreach (JsonValue input in recipe["inputs"].Items) state.Player.Inventory.Remove(input["itemId"].AsString(), (int)input["qty"].AsNumber());
            string output = recipe["output"]["itemId"].AsString();
            int amount = (int)recipe["output"]["qty"].AsNumber(1);
            if (!burned && state.Player.Inventory.IsBulkItem(output) && state.Player.Inventory.StoredAmount() + amount > state.Player.Inventory.StorageCap) { Stop(CraftEndReason.InventoryFull); return; }
            int xp = burned ? (int)Math.Round(recipe["xp"].AsNumber() * .2) : (int)recipe["xp"].AsNumber();
            if (!burned) { state.Player.Inventory.Add(output, amount); state.CollectionLog.Add(output); } else amount = 0;
            state.Player.Skills.AddXp(skill, xp);
            state.Player.Skills.AddMasteryXp(skill, recipe["id"].AsString(), Math.Max(1, (int)recipe["output"]["qty"].AsNumber(1)));
            Crafted?.Invoke(new CraftEvent { Recipe = recipe, Amount = amount, XpGained = xp, Preserved = preserved, Burned = burned });
            CraftEndReason? next = CanStart(recipe);
            if (next.HasValue) Stop(next.Value);
        }
    }
}
